### DISCUSSION TITLE
Typescript: Session Types

### DISCUSSION THREAD

**Discussion by cdt-eth** [supabase/supabase#2222] [Questions]
Hey!

I'm refactoring an app to Typescript and I'm fairly new to TS and unsure how to handle these 2 errors. Any help or clarification would be appreciated!

```typescript
const [session, setSession] = useState(null);
const user = supabase.auth.user();

useEffect(() => {
  setSession(supabase.auth.session()); 

  supabase.auth.onAuthStateChange((_event, session) => {
    setSession(session);
  });
}, []);
```

1) "Argument of type 'Session | null' is not assignable to parameter of type 'SetStateAction<null>'." 
`setSession(supabase.auth.session());`

2) "Argument of type 'Session | null' is not assignable to parameter of type 'SetStateAction<null>'.ts"
`setSession(session);`

**Comment by borngraced** [comment]
I'd suggest you first console.log the data returned from session and then proceed by creating a type/interface for it.

after doing so, use it like this `const [session, setSession] = useState<NewlyCreatedTypeName | null>(null);`

Let me if that works. thanks

**Reply by cdt-eth** [reply]
Hey! So I console.log'd the data and its an object, but I tried assigning `session`, `Session`, `user`, and `supabase` to `object`, in different combinations, but it always gave an error. 

Once I deleted them all, and had an empty interface, it worked, but I don't understand why. I'm not sure why it works when empty and what it should be filled in with.

This is all I did and it compiled successfully:
```typescript
interface INav {}

const Nav = (): ReactElement => {
  const [session, setSession] = useState<INav | null>(null);
  . 
  .
  .
}
```

**Comment by jonohewitt** [comment]
You could use the Session type provided with supabase:

```
import { Session } from "@supabase/gotrue-js/src/lib/types"

...

const [session, setSession] = useState<Session>()

```


To find this you can hover in VS Code, e.g:

<img width="506" alt="Screenshot 2021-07-07 at 12 33 10" src="https://user-images.githubusercontent.com/47791294/124745128-c363a600-df1f-11eb-80ee-1f2f83c58c68.png">

Then you can just start typing the name in and it can be auto-imported, e.g:

<img width="804" alt="Screenshot 2021-07-07 at 12 45 07" src="https://user-images.githubusercontent.com/47791294/124746540-46d1c700-df21-11eb-942c-18c4090ee719.png">


You can inspect the type definition with Right click > Go to Type Definition 

// GoTrueClient.ts

<img width="527" alt="Screenshot 2021-07-07 at 12 36 58" src="https://user-images.githubusercontent.com/47791294/124745370-0887d800-df20-11eb-817c-c29df3943dfe.png">

Then Right click > Go to Type Definition again on the Session interface 

//types.ts

<img width="641" alt="Screenshot 2021-07-07 at 12 37 50" src="https://user-images.githubusercontent.com/47791294/124745477-26553d00-df20-11eb-9806-792da87bcf7f.png">

**Reply by cdt-eth** [reply]
thanks! that helped for that file.

How would I find the types for the `Account` component? It's taken directly from [Supabase's React boilerplate](https://supabase.io/docs/guides/with-react#account-page). I found the `types.ts` file but see nothing for `username`, `website`, and `avatar_url`

**Reply by jonohewitt** [reply]
Those just look like basic strings so they don't need a custom type definition. You can use the built-in `string` type, or let typescript infer it automatically by assigning a string as the default value.

Maybe I'm missing something but that's probably what I would have done by using empty quotes instead of null as the default, e.g instead of
```
const [username, setUsername] = useState(null) 

// username has the 'any' type

...

value={username || ""}
```
I would write
```
const [username, setUsername] = useState("") 

// username is automatically given the 'string' type

...

value={username}
```

But you could also keep the original code and explicitly specify the string type
```
const [username, setUsername] = useState<string>(null) 

// username is explicitly typed as 'string'

...

value={username || ""}
```

**Reply by BenGardiner123** [reply]
I just wanted to say a big thank you to @jonohewitt and to people like you that show patience and give answers like this. 
Thank you man.

**Reply by jonohewitt** [reply]
Hey @BenGardiner123 sorry for the slow reply but thanks a lot for the comment! Really appreciate it and glad my posts could help out.

**Comment by cdt-eth** [comment]
Thanks, that makes a lot of sense. Feels silly to have asked now, that one was easy.

I have a final question, if you don't mind.

in `Auth.tsx` I have this code and and it passes `session` as a prop:
<img width="637" alt="Screen Shot 2021-07-08 at 9 05 09 AM" src="https://user-images.githubusercontent.com/26611339/124926514-9c62ae00-dfcb-11eb-8fea-88fa1a2837af.png">

in `Account.tsx` I receive the session, but it says `Binding element 'session' implicitly has an 'any'`, how do I type this? For `user` below it says `Object is possibly 'null'`:
<img width="636" alt="Screen Shot 2021-07-08 at 9 06 31 AM" src="https://user-images.githubusercontent.com/26611339/124926657-cb791f80-dfcb-11eb-8593-211c136ec9ba.png">

**Reply by jonohewitt** [reply]
No worries! 

For functions you can either assign types inline, usually if you only have a small number of props and you won't be reusing the definition elsewhere, or create a custom definition. So you could do:

```
    const Account = ({ session }: { session: Session }) => {}
```
or
```
    interface Account {
      session: Session
    }

    const Account = ({ session }: Account) => {}
```
There's also another form using `type` instead of `interface`
```
    type Account = {
      session: Session
    }

    const Account = ({ session }: Account) => {}
```
[Type aliases](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-aliases) are slightly different from [interfaces](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#interfaces) and although it would also work fine here, it's recommended to stick to `interface` unless you need the extra functionality. 

If you haven't seen it yet, it's definitely worth checking out the [Typescript Handbook](https://www.typescriptlang.org/docs/handbook/2/basic-types.html)

**Reply by cdt-eth** [reply]
Okay great, that makes sense. and when  it says "Object is possibly 'null'" i'd add a `?` to the property correct? 

```typescript
const updates = {
        id: user?.id,
        .
        .
}

let { data, error, status } = await supabase
        .from("profiles")
        .select(`username, website, avatar_url`)
        .eq("id", user?.id)
        .single();
```

**Reply by jonohewitt** [reply]
The question mark there is called [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) which is just normal javascript - basically it prevents a breaking crash error if the object isn't found and returns `undefined` instead - which probably isn't what you want to do there.

 I would make sure that the user id exists before making that fetch, maybe wrapping it in an if statement or putting it in a function and only calling it somewhere you've confirmed the user id exists.

**Reply by datalifenyc** [reply]
> Thanks, that makes a lot of sense. Feels silly to have asked now, that one was easy.
> 
> I have a final question, if you don't mind.
> 
> in `Auth.tsx` I have this code and and it passes `session` as a prop: <img alt="Screen Shot 2021-07-08 at 9 05 09 AM" width="637" src="https://user-images.githubusercontent.com/26611339/124926514-9c62ae00-dfcb-11eb-8fea-88fa1a2837af.png">
> 
> in `Account.tsx` I receive the session, but it says `Binding element 'session' implicitly has an 'any'`, how do I type this? For `user` below it says `Object is possibly 'null'`: <img alt="Screen Shot 2021-07-08 at 9 06 31 AM" width="636" src="https://user-images.githubusercontent.com/26611339/124926657-cb791f80-dfcb-11eb-8593-211c136ec9ba.png">

This pointed me in the right direction.

authStore.js
[Note: `User` can be replaced with `Session`]

```js
import { writable } from 'svelte/store';
import { User } from "@supabase/gotrue-js/src/lib/types"

export const user = writable(User())
```

__layout.svelte

```html
<script>
    import { supabase } from "../supabase.js";
    import { user } from "../stores/authStore.js";

    user.set(supabase.auth.user());
</script>
```

**Reply by clodal** [reply]
In the latest version as of Jan 2025, the import path is now at 

`import { User } from '@supabase/auth-js/src/lib/types'`

Update: I don't recommend importing from src, it'll cause compilation errors at build
